ML project/Customer_churn_prediction.py

```diff
@@ -0,0 +1,512 @@
+
+# In[11]:
+
+
+cols= list(df.select_dtypes(['int64','float64']).columns)
+
+upper_limit = df[cols].quantile(0.75) + 1.5*(df[cols].quantile(0.75)-df[cols].quantile(0.25))
+lower_limit = df[cols].quantile(0.25) - 1.5*(df[cols].quantile(0.75)-df[cols].quantile(0.25))
+cleaned_df = df[~((df[cols]>upper_limit)|(df[cols]<lower_limit)).any(axis=1)]
+cleaned_df.shape
+
+
+# In[12]:
+
+
+print('Percentage of removed data: {:.2f}'.format(100-cleaned_df.shape[0]/df.shape[0]*100))
+
+
+# In[13]:
+
+
+for i in cleaned_df.select_dtypes(['int64','float64']).columns:
+    sns.histplot(cleaned_df[i]);
+    plt.show();
+
+
+# The majority of numerical features shows a gaussian distribution, hence StandardScaler could be used too. Let's try MinMaxScaler
+
+# In[14]:
+
+
+from sklearn.preprocessing import MinMaxScaler
+scaler = MinMaxScaler()
+df_scaled = cleaned_df.copy()
+df_scaled[df_scaled.select_dtypes(['int64','float64']).columns] = scaler.fit_transform(df_scaled[df_scaled.select_dtypes(['int64','float64']).columns])
+
+
+# Let's format bynary categorical features and target column as needed
+
+# In[15]:
+
+
+df_scaled['international_plan'] = df_scaled.international_plan.apply(lambda x: 1 if x == 'yes' else 0)
+df_scaled['voice_mail_plan'] = df_scaled.voice_mail_plan.apply(lambda x: 1 if x == 'yes' else 0)
+df_scaled['churn'] = df_scaled.churn.apply(lambda x: 1 if x == 'yes' else 0)
+
+
+# In[16]:
+
+
+df_scaled = pd.get_dummies(df_scaled)
+
+
+# In[17]:
+
+
+df_scaled.churn.value_counts(normalize=True)
+
+
+# The target column shows that the dataset is unbalanced, hence stratifying based on this column would be a good choice
+
+# ### KNearestNeighbors
+
+# In[18]:
+
+
+from sklearn.neighbors import KNeighborsClassifier
+from sklearn.model_selection import train_test_split
+from sklearn.metrics import accuracy_score, roc_curve, roc_auc_score, f1_score, balanced_accuracy_score, precision_recall_curve
+from sklearn.model_selection import RandomizedSearchCV, cross_val_score
+
+
+# In[19]:
+
+
+X = df_scaled.drop('churn',axis=1).values
+y = df_scaled.churn.values
+
+
+# In[20]:
+
+
+X_train, X_test, y_train, y_test = train_test_split(X,y,test_size=0.2,stratify=y)
+
+
+# In[21]:
+
+
+params = {'n_neighbors':[i for i in range(1,11)]}
+knc = KNeighborsClassifier()
+for i in ['f1','balanced_accuracy','accuracy','roc_auc']:
+    rcv = RandomizedSearchCV(knc, param_distributions = params, cv=3,scoring = i,random_state=123)
+    rcv.fit(X_train,y_train)
+    print('Metrics: {}\nBest params: {}\nBest score: {}'.format(i,rcv.best_params_,rcv.best_score_))
+    
+
+
+# In[22]:
+
+
+knc_f1 = []
+knc_b_acc = []
+knc_acc = []
+knc_auc = []
+
+for i in ['f1','balanced_accuracy','accuracy','roc_auc']:
+    knc = KNeighborsClassifier(n_neighbors=10)
+    metrics = cross_val_score(knc, X_train, y_train, cv = 10, scoring=i)
```

Comment:
@Ettore1991 with this code metrics run with different train/test split. Not 100% fair to compare